vm/sdk/wallet/tx.go

```diff
@@ -25,61 +28,77 @@ func encode(fields ...scale.Encodable) []byte {
 
-	public := &core.PublicKey{}
-	copy(public[:], signing.Public(pk))
 	// note that principal is computed from pk
-	principal := core.ComputePrincipal(wallet.TemplateAddress, public)
+	athenaPayload := vmlib.EncodeTxSpawn(athcon.Bytes32(signing.Public(pk)))
+	principal, err := core.ComputePrincipalFromPubkey(wallet.TemplateAddress, *signing.NewPublicKey(signing.Public(pk)))
+	if err != nil {
+		return nil, err
+	}
+	payload := core.Payload(athenaPayload)
 
-	// TODO(lane): fix encoding
-	tx := encode(&sdk.TxVersion, &principal, &template, &payload, args)
-	sig := ed25519.Sign(ed25519.PrivateKey(pk), core.SigningBody(options.GenesisID[:], tx))
-	return append(tx, sig...)
+	// The payload is already encoded. Why, might you ask, are we encoding it again?
+	// Short answer: because, when decoding txs, go-spacemesh can only decode SCALE-encoded data.
+	// Fixing this, and allowing a tx to be partially SCALE-encoded, partially raw bytes,
+	// is a lot of work for a tiny bit of gain.
+	tx := encode(&sdk.TxVersion, &principal, &meta, &payload)
+	// tx := encode(&sdk.TxVersion, &principal, &meta)
+	// tx = append(tx, payload...)
+
+	// sig := ed25519.Sign(ed25519.PrivateKey(pk), core.SigningBody(options.GenesisID[:], tx))
+	sig := ed25519.Sign(ed25519.PrivateKey(pk), tx)
+	return append(tx, sig...), nil
 }
 
 // Spend creates a spend transaction.
-func Spend(pk signing.PrivateKey, to types.Address, amount uint64, nonce types.Nonce, opts ...sdk.Opt) []byte {
+func Spend(pk signing.PrivateKey, to types.Address, amount uint64, nonce types.Nonce, opts ...sdk.Opt) ([]byte, error) {
 	options := sdk.Defaults()
 	for _, opt := range opts {
 		opt(options)
 	}
 
-	spawnargs := wallet.SpawnArguments{}
-	copy(spawnargs.PublicKey[:], signing.Public(pk))
-	principal := core.ComputePrincipal(wallet.TemplateAddress, &spawnargs)
+	// Encode using the VM
+	vmlib, err := athcon.LoadLibrary(host.AthenaLibPath())
+	if err != nil {
+		panic(fmt.Errorf("loading Athena VM: %w", err))
+	}
+
+	principal, err := core.ComputePrincipalFromPubkey(wallet.TemplateAddress, *signing.NewPublicKey(signing.Public(pk)))
+	if err != nil {
+		return nil, err
+	}
+
+	payload := core.Payload(vmlib.EncodeTxSpend(athcon.Address(to), amount))
 
-	payload := core.Payload{}
-	payload.GasPrice = options.GasPrice
-	payload.Nonce = nonce
+	meta := core.Metadata{}
+	meta.GasPrice = options.GasPrice
+	meta.Nonce = nonce
 
-	args := wallet.SpendArguments{}
-	args.Destination = to
-	args.Amount = amount
+	tx := encode(&sdk.TxVersion, &principal, &meta, &payload)
+	// tx := encode(&sdk.TxVersion, &principal, &meta)
+	// tx = append(tx, payload...)
 
-	// TODO(lane): fix encoding
-	tx := encode(&sdk.TxVersion, &principal, &payload, &args)
-	sig := ed25519.Sign(ed25519.PrivateKey(pk), core.SigningBody(options.GenesisID[:], tx))
-	return append(tx, sig...)
+	// sig := ed25519.Sign(ed25519.PrivateKey(pk), core.SigningBody(options.GenesisID[:], tx))
+	sig := ed25519.Sign(ed25519.PrivateKey(pk), tx)
```

Comment:
What about this?

Reply:
not sure what you mean

Reply:
Sorry, I meant that there is  `// sig := ed25519.Sign(ed25519.PrivateKey(pk), core.SigningBody(options.GenesisID[:], tx))` left. I understand that genesis should be included - I'm asking what the plan is for it. If not, the commented line could be removed.

Reply:
We need to decide how we want to implement genesisID support in Athena. Assuming we do it in a way that's compatible with what go-spacemesh does today, then these commented lines can just be restored once that's been added on the Athena side.